lib/auth/ec2_join.go

```diff
@@ -0,0 +1,861 @@
+	var err error
+
+	switch req.Role {
+	case types.RoleNode:
+		instanceExists, err = nodeExists(ctx, a, req.HostID)
+	case types.RoleProxy:
+		instanceExists, err = proxyExists(ctx, a, req.HostID)
+	case types.RoleKube:
+		instanceExists, err = kubeExists(ctx, a, req.HostID)
+	case types.RoleApp:
+		instanceExists, err = appExists(ctx, a, req.HostID)
+	case types.RoleDatabase:
+		instanceExists, err = dbExists(ctx, a, req.HostID)
+	default:
+		return trace.BadParameter("unsupported role: %q", req.Role)
+	}
+
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if instanceExists {
+		log.Warnf("Server with ID %q and role %q is attempting to join the cluster with a Simplified Node Joining request, but"+
+			" a server with this ID is already present in the cluster.", req.HostID, req.Role)
+		return trace.AccessDenied("server with host ID %q and role %q already exists", req.HostID, req.Role)
+	}
+	return nil
+}
+
+// CheckEC2Request checks register requests which use EC2 Simplified Node
+// Joining. This method checks that:
+// 1. The given Instance Identity Document has a valid signature (signed by AWS).
+// 2. A node has not already joined the cluster from this EC2 instance (to
+//    prevent re-use of a stolen Instance Identity Document).
+// 3. The signed instance attributes match one of the allow rules for the
+//    corresponding token.
+// If the request does not include an Instance Identity Document, and the
+// token does not include any allow rules, this method returns nil and the
+// normal token checking logic resumes.
+func (a *Server) CheckEC2Request(ctx context.Context, req RegisterUsingTokenRequest) error {
+	requestIncludesIID := req.EC2IdentityDocument != nil
+	token, err := a.GetCache().GetToken(ctx, req.Token)
+	if err != nil {
+		if trace.IsNotFound(err) && !requestIncludesIID {
+			// This is not a Simplified Node Joining request, pass on to the
+			// regular token checking logic in case this is a static token.
+			return nil
+		}
+		return trace.Wrap(err)
+	}
+	tokenRequiresIID := len(token.GetAllowRules()) > 0
+
+	if !requestIncludesIID && !tokenRequiresIID {
+		// not a simplified node joining request, pass on to the regular token
+		// checking logic
+		return nil
+	}
+	if tokenRequiresIID && !requestIncludesIID {
+		return trace.AccessDenied("this token requires an EC2 Identity Document from the node")
+	}
+	if !tokenRequiresIID && requestIncludesIID {
+		return trace.BadParameter("an EC2 Identity Document is included in a register request for a token which does not expect it")
+	}
+
+	iid, err := parseAndVerifyIID(req.EC2IdentityDocument)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := checkPendingTime(iid, a.clock); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := a.checkInstanceUnique(ctx, req, iid); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := checkEC2AllowRules(ctx, iid, token); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+var awsCertBytes = map[string][]byte{
```

Comment:
for readability can we move it to a separate file ?

Reply:
yes, good idea, done